.mergify.yml

```diff
@@ -0,0 +1,39 @@
+pull_request_rules:
+  - name: backport patches to 7.x branch
+    conditions:
+      - base=master
+      - label=v7.13.0
+    actions:
+      backport:
+        branches:
+          - "7.x"
+  - name: backport patches to 7.12 branch
+    conditions:
+      - base=master
+      - label=v7.12.0
```

Comment:
How exactly will the labels work?

a. If we are about to release 7.12.0, I might want to use the label `v7.12.1`, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

b. Will my PR be backported if I add `v7.12.0` even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

Reply:
I'll answer your questions in a minute, for clarity, the initial proposal was to use the existing labels `vmajor.minor.0` (see https://github.com/elastic/beats/pull/24576#issuecomment-801099897) 

> a. If we are about to release 7.12.0, I might want to use the label v7.12.1, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

Yes, `mergify` works for even PRs which were not backported after the merge, as long as the label is added. In addition you can use the `github comment` = `@Mergifyio backport <destination> will backport this PR on <destination> branch`

To illustrate the above, I just tagged https://github.com/elastic/apm-pipeline-library/pull/1053 with the explicit label and https://github.com/elastic/apm-pipeline-library/pull/1056 was created. NOTE: for some reason the timestamp for the label does not match with when I did it, (about 7 minutes ago instead of 2 days...)

> b. Will my PR be backported if I add v7.12.0 even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

Yes, backports don't know about releases but branches and tags.

Reply:
>a. If we are about to release 7.12.0, I might want to use the label v7.12.1, in order to postpone the backport for after the release is out. Will we update the version filter here after a release? If so, will my PR automatically backported when we update the label filter here?

if the mergify config is updated with that branch before you merge the PR, the PR would backported.

>b. Will my PR be backported if I add v7.12.0 even after lets say 7.12.1 was released? (Although it might be a bit confusing to users looking at the PR labels to figure out if a bug fix is available)

the backport is made on a new PR when you merge the fixed PR, if the labels and mergify configuration are in place when you merge, a new PR is made to every branch that you requested backport with the labels.

There is an alternative way to make the backports manually using a GitHub comment (see  https://docs.mergify.io/commands.html#backport)

https://mergify.io/features/backports

Reply:
Maybe to clarify case a:
1. mergify only backports if label `v7.12.0` is set
2. I create a PR and merge it
3. I add the label `v7.12.1` to my PR (because of FF I don't want the change be backported yet)
4. Version 7.12.0 gets released
5. After 7.12.0 is released we update the mergify configuration to backport if `v7.12.0` or `v7.12.1` is set.

After step 5, will my old PR be picked up automatically or am I required to remember that I did have a 2 weeks old PR that I must now 'manually' backport by adding a comment to the PR (or remove + add label again).

Case b is a little unfortunate as labels do not necesarrily reflect reality, but well, it's the way it is.

Reply:
>After 7.12.0 is released we update the mergify configuration to backport if v7.12.0 or v7.12.1 is set.

this happens after you merge the PR, mergify does not review old merged PRs, so the backport is not made. In this case, you have to manually backport the PR, either from the PR with a GiHub Comment or as you make now from your laptop.
However, if we detect changes on the mergify configfile can trigger something that reviews all PRs with the new label and make backport mergify comments on those PRs, but this is something not out of the box. We have to see if this use case exists and how many times it happens, then we can evaluate the effort and the value of this process.

Reply:
AFAIK if you fix something in branch 7.12 that has the version 7.12.0 this will be in the 7.12 (the same major.minor different patch) that will contain the version 7.12.1, so in the case you put as an example there is no backport.
I can see a case similar of what you explain between minor versions, but in those case you use the branch `7.x` and then you backport to `7.12`, `7.11`, ... in those cases the branch already exists so the backport is made.

Reply:
To help to unblock this 🧵 

Since there is no backport automation in place for this project I see the benefits of enabling it for facilitating the backporting process, but I don't know if it could cause any issues. Although, `mergify` requires certain discipline regarding what labels should be added. 

Do you think if we enable `mergify` it will help with the backporting? Or is it required to solve all of the above-mentioned use cases?

Reply:
I don't think either behavior is a blocker and adding mergify is an improvement either way. As we have to document the changelog process in our how to guides. I'm trying to figure out how it works and where we have potential gotchas that we should document as well. I did just bring up the two cases as we've seen them from time to time in the past: label does not match actual release + us holding of with backport in order to target the next patch release... cough... kafka ... cough.